File: services/exchange_rate.py
"""
為替レート取得サービス。
無料API（open.er-api.com）を使用。APIキー不要。
"""
import requests
from utils.cache_manager import get_exchange_rate_from_cache, save_exchange_rate_to_cache
from config.settings import AppConfig


def get_cny_to_jpy() -> float:
    """人民元→円の為替レートを返す。キャッシュ優先。"""
    if AppConfig.TEST_MODE:
        return 21.56  # テストモード固定値

    pair = "CNY_JPY"
    cached = get_exchange_rate_from_cache(pair)
    if cached is not None:
        return cached

    rate = _fetch_from_api()
    save_exchange_rate_to_cache(pair, rate)
    return rate


def _fetch_from_api() -> float:
    """外部APIから人民元→円レートを取得する。"""
    try:
        # open.er-api.com: 無料・APIキー不要
        resp = requests.get(
            "https://open.er-api.com/v6/latest/CNY",
            timeout=10,
        )
        resp.raise_for_status()
        data = resp.json()
        rate = data["rates"]["JPY"]
        return float(rate)
    except Exception:
        pass

    try:
        # フォールバック: frankfurter.app
        resp = requests.get(
            "https://api.frankfurter.app/latest?from=CNY&to=JPY",
            timeout=10,
        )
        resp.raise_for_status()
        data = resp.json()
        return float(data["rates"]["JPY"])
    except Exception:
        # 最終フォールバック: 固定値
        return 21.5

File: netlify/functions/exchange-rate.mjs
export async function handler(event) {
  const headers = {
    "Content-Type": "application/json",
    "Access-Control-Allow-Origin": "*",
  };

  if (event.httpMethod === "OPTIONS") {
    return { statusCode: 204, headers };
  }

  try {
    // open.er-api.com（無料・APIキー不要）
    let rate = null;
    try {
      const res = await fetch("https://open.er-api.com/v6/latest/CNY", { signal: AbortSignal.timeout(8000) });
      const data = await res.json();
      rate = data.rates?.JPY;
    } catch {}

    // フォールバック: frankfurter.app
    if (!rate) {
      try {
        const res = await fetch("https://api.frankfurter.app/latest?from=CNY&to=JPY", { signal: AbortSignal.timeout(8000) });
        const data = await res.json();
        rate = data.rates?.JPY;
      } catch {}
    }

    // 最終フォールバック
    if (!rate) rate = 21.5;

    return {
      statusCode: 200,
      headers,
      body: JSON.stringify({ rate, pair: "CNY_JPY" }),
    };
  } catch (err) {
    return {
      statusCode: 500,
      headers,
      body: JSON.stringify({ error: err.message }),
    };
  }
}
